lib/spack/spack/ci/__init__.py

```diff
@@ -465,28 +480,51 @@ def generate_pipeline(env: ev.Environment, args) -> None:
         # pruning.  Otherwise, list the names of all packages touched between
         # rev1 and rev2, and prune from the pipeline any node whose spec has a
         # packagen name not in that list.
-        rev1, rev2 = get_change_revisions()
-        tty.debug(f"Got following revisions: rev1={rev1}, rev2={rev2}")
+
+        # Check if the environment changed
+        stack_changed = False
+        rev1, rev2 = get_change_revisions(os.path.dirname(env.manifest_path))
         if rev1 and rev2:
-            # If the stack file itself did not change, proceed with pruning
-            if not get_stack_changed(env.manifest_path, rev1, rev2):
-                affected_pkgs = compute_affected_packages(rev1, rev2)
-                tty.debug("affected pkgs:")
-                for p in affected_pkgs:
-                    tty.debug(f"  {p}")
-                affected_specs = get_spec_filter_list(
-                    env,
-                    affected_pkgs,
-                    dependent_traverse_depth=options.untouched_pruning_dependent_depth,
-                )
-                tty.debug(
-                    "dependent_traverse_depth="
-                    f"{options.untouched_pruning_dependent_depth}, affected specs:"
-                )
-                for s in affected_specs:
-                    tty.debug(f"  {PipelineDag.key(s)}")
-
-                pruning_filters.append(create_unaffected_pruner(affected_specs))
+            stack_changed = get_stack_changed(env.manifest_path, rev1, rev2)
+
+        # If the stack env has changed, do not apply unaffected pruning
+        if not stack_changed:
+            # TODO: This should be configurable to only check for changed packages
+            # in specific configured repos that are being tested with CI. For now
+            # it assumes all configured repos are merge commits that contain relevant
+            # changes to run CI on.
+            for repo in spack.repo.PATH.repos:
+                # Use the dirname of the repo.root to cache repos under the same prefix
+                rev1, rev2 = get_change_revisions(os.path.dirname(repo.root))
+                tty.debug(f"repo {repo.namespace}: revisions rev1={rev1}, rev2={rev2}")
+                if rev1 and rev2:
+                    affected_pkgs = compute_affected_packages(repo, rev1, rev2)
+                    tty.debug(f"repo {repo.namespace}: affected pkgs")
+                    for p in affected_pkgs:
+                        tty.debug(f"  {p}")
+                    affected_specs = get_spec_filter_list(
+                        env,
+                        affected_pkgs,
+                        dependent_traverse_depth=options.untouched_pruning_dependent_depth,
+                    )
```

Comment:
This should probably not be in the loop over repos.

Given you're not testing the multiple package repos case I would suggest to not generalize the functionality and stick to `builtin` only.

Reply:
Added a test for this.